### ISSUE TITLE
Is there TensorFlow implementation of  vocoder(waveRNN)?

### ISSUE THREAD

**Issue by ymzlygw** [CorentinJ/Real-Time-Voice-Cloning#843]
Hi , I want to know that Is there TensorFlow implementation of  vocoder(waveRNN)? 
I ask this because that when I try to change the Pytorch model of waveRNN to tensorflow 2.x version, I succeed and can train it on dataset, but I find that the inference speed is 10 times slow that waveRNN of pytorch....
### It occurs in 'for loop' of  [vocoder.generate()] function:
def generate(self, mels, batched, target, overlap, mu_law, progress_callback=None):
................

     for i in range(seq_len):

                m_t = mels[:, i, :]
                a1_t, a2_t, a3_t, a4_t = (a[:, i, :] for a in aux_split)

                x = torch.cat([x, m_t, a1_t], dim=1)
                x = self.I(x)
                h1 = rnn1(x, h1)

                x = x + h1
                inp = torch.cat([x, a2_t], dim=1)

                h2 = rnn2(inp, h2)

                x = x + h2
                x = torch.cat([x, a3_t], dim=1)
                x = F.relu(self.fc1(x))

                x = torch.cat([x, a4_t], dim=1)
                x = F.relu(self.fc2(x))

                logits = self.fc3(x)

                if self.mode == 'MOL':
                    sample = sample_from_discretized_mix_logistic(logits.unsqueeze(0).transpose(1, 2))
                    output.append(sample.view(-1))
                    if torch.cuda.is_available():
                        # x = torch.FloatTensor([[sample]]).cuda()
                        x = sample.transpose(0, 1).cuda()
                    else:
                        x = sample.transpose(0, 1)

                elif self.mode == 'RAW' :
                    posterior = F.softmax(logits, dim=1)
                    distrib = torch.distributions.Categorical(posterior)

                    sample = 2 * distrib.sample().float() / (self.n_classes - 1.) - 1.
                    output.append(sample)
                    x = sample.unsqueeze(-1)
                else:
                    raise RuntimeError("Unknown model mode value - ", self.mode)

                if i % 100 == 0:
                    gen_rate = (i + 1) / (time.time() - start) * b_size / 1000
                    progress_callback(i, seq_len, b_size, gen_rate)
.............
### Tensorflow2.3 version:

      for i in range(seq_len):

            m_t = mels[:, i, :]

            a1_t, a2_t, a3_t, a4_t = (a[:, i, :] for a in aux_split)

            x = tf.concat([x, m_t, a1_t], axis=1)
            x = self.I(x)

            _, h1 = rnn1(tf.expand_dims(x, axis=1))

            x = x + h1
            inp = tf.concat([x, a2_t], axis=1)
            _, h2 = rnn2(tf.expand_dims(inp, axis=1))

            x = x + h2
            x = tf.concat([x, a3_t], axis=1)
            x = tf.nn.relu(self.fc1(x))

            x = tf.concat([x, a4_t], axis=1)
            x = tf.nn.relu(self.fc2(x))

            logits = self.fc3(x)

            if self.mode == 'RAW' :
                posterior = tf.nn.softmax(logits, axis=1)
                distrib = tfp.distributions.Categorical(posterior, dtype=tf.float32)
             
                sample = 2 * distrib.sample() / (self.n_classes - 1.) - 1.
                output.append(sample)
                x = tf.expand_dims(sample, axis=-1)
            else:
                raise RuntimeError("Unknown model mode value - ", self.mode)

            if i % 100 == 0:
                gen_rate = (i + 1) / (time.time() - start) * b_size_np / 1000
                progress_callback(i, seq_len, b_size, gen_rate)
...............

#### This part generate 1 sample using 5s for pytorch, but even 50s in tensorflow2.3! And I don't know why.
Could you please help me with the problem?

**Comment by ymzlygw** [comment]
Anyone can help me ?

**Comment by blue-fish** [comment]
This is a Pytorch repo, questions about Tensorflow are out of scope unfortunately. You can try asking your question in the fatchord/WaveRNN repo.
